docs/content/docs/connectors/datastream/datagen.md

```diff
@@ -0,0 +1,115 @@
+specific language governing permissions and limitations
+under the License.
+-->
+
+# DataGen Connector
+
+The DataGen connector provides a `Source` implementation that allows for generating input data for 
+Flink pipelines.
+It is useful when developing locally or demoing without access to external systems such as Kafka.
+The DataGen connector is built-in, no additional dependencies are required.
+
+Usage
+-----
+
+The `DataGeneratorSource` produces N data points in parallel. The source splits the sequence 
+into as many parallel sub-sequences as there are parallel source subtasks. It drives the data 
+generation process by supplying "index" values of type Long to the user-provided 
+{{< javadoc name="GeneratorFunction" file="org/apache/flink/connector/datagen/source/GeneratorFunction.html" >}}.
+
+The `GeneratorFunction` is then used for mapping the (sub-)sequences of Long values
+into the generated events of an arbitrary data type. For instance, the following code will produce the sequence of
+`["Number: 0", "Number: 2", ... , "Number: 999"]` records.
+
+```java
+GeneratorFunction<Long, String> generatorFunction = index -> "Number: " + index;
+
+DataGeneratorSource<String> source =
+        new DataGeneratorSource<>(generatorFunction, 1000, Types.STRING);
+
+DataStreamSource<String> stream =
+        env.fromSource(source,
+        WatermarkStrategy.noWatermarks(),
+        "Generator Source");
+```
+
+The order of elements depends on the parallelism. Each sub-sequence will be produced in order.
+Consequently, if the parallelism is limited to one, this will produce one sequence in order from
+`"Number: 0"` to `"Number: 999"`.
+
+`DataGeneratorSource` has built-in support for rate limiting. The following code will produce an
+effectively unbounded (`Long.MAX_VALUE` from a practical perspective will never be reached) stream of
+Long values at the overall source rate (across all source subtasks) not exceeding 100 events per second.
+
+```java
+GeneratorFunction<Long, Long> generatorFunction = index -> index;
+
+DataGeneratorSource<String> source =
+        new DataGeneratorSource<>(
+             generatorFunctionStateless,
+             Long.MAX_VALUE,
+             RateLimiterStrategy.perSecond(100),
+             Types.STRING);
+```
+
+The source also allows for producing specific elements between the checkpoint boundaries using the 
+corresponding 
+{{< javadoc name="RateLimiterStrategy" file="org/apache/flink/api/connector/source/util/ratelimit/RateLimiterStrategy.html">}}. 
+This is particularly useful for testing scenarios where certain output
+is expected to be produced upon checkpoint completions. The below snippet illustrates an example of 
+producing the sequence of elements `"a","b", .. ,"j"` repeatedly between checkpoints:
+
+```java
+StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+env.enableCheckpointing(3000);
+env.setRuntimeMode(RuntimeExecutionMode.STREAMING);
```

Comment:
The underlying `NumberSequenceSource` has this bit:
```
This source is always bounded. For very long sequences (for example over the entire domain of
long integer values), user may want to consider executing the application in a streaming manner,
because, despite the fact that the produced stream is bounded, the end bound is pretty far away.
```
We could either emphasize this or just skip mentioning it altogether and also remove the settings (since it is the default). Since I am coming completely from the "streaming perspective", I am not sure how many people would want to use this source in their batch jobs.

Reply:
Added a separate Boudedness section.